src/parser.cpp

```diff
@@ -249,27 +248,22 @@ static void append_block_description_to_stack_trace(const parser_t &parser, cons
                 // TODO: Escape these.
                 append_format(trace, _(L" with arguments '%ls'"), args_str.c_str());
             }
-            trace.push_back('\n');
-            print_call_site = true;
             break;
         }
         case block_type_t::subst: {
-            append_format(trace, _(L"in command substitution\n"));
-            print_call_site = true;
+            append_format(trace, _(L"in command substitution"));
             break;
         }
         case block_type_t::source: {
             const filename_ref_t &source_dest = b.sourced_file;
-            append_format(trace, _(L"from sourcing file %ls\n"),
+            append_format(trace, _(L"from sourcing file %ls"),
                           user_presentable_path(*source_dest, parser.vars()).c_str());
-            print_call_site = true;
             break;
         }
         case block_type_t::event: {
             assert(b.event && "Should have an event");
             wcstring description = *event_get_desc(parser, **b.event);
-            append_format(trace, _(L"in event handler: %ls\n"), description.c_str());
-            print_call_site = true;
+            append_format(trace, _(L"in event handler: %ls"), description.c_str());
```

Comment:
The only potential downside is it could look worse if the argument strings are overlong.
I think that's an edge case so probably this is fine and overall a great change.

---

As a programmer, the stack traces are a bit confusing to me, if I see something like

	function foo file bar.fish:3
	function bar ...

I expect that line 3 in bar.fish is inside function `foo`. But that's a separate concern

Also I expect the ordering file,line,function,arguments  instead of function,arguments,file,line (which would alleviate the concern about long arguments)

---

> I would be entirely fine if we decided to go with just the first one, or if we decided to put the rest off for 4.0, or if the rest gets in the way of someone's porting and I need to write it again later, or if we decide to go for a cleaner solution.

I don't mind merging things like this first (it's small and I don't expect too many of them).

Reply:
> The only potential downside is it could look worse if the argument strings are overlong.

There are a bunch of ways to solve this:

1. Truncate/ellipsize the arguments (not great)
2. Move the arguments onto another line if they are very long

Something like

```
fish: Unknown command: nonexistent
foo.fish (line 4): 
    if nonexistent
       ^~~~~~~~~~^
in function 'this-errors' with arguments:
'foo\ bar aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa baz'
 called on line 11 of file foo.fish
in function 'call-this' called on line 1 of file foo.fish
```

for the latter. I have something whipped up, I'll add it here.

>Also I expect the ordering file,line,function,arguments instead of function,arguments,file,line

I think that would be awkward because we don't tend to use tons of files. So you would have a lot of

```
foo.fish:50:foo bar baz
foo.fish:42:function2 arguments
foo.fish:38:function1 arguments
```

which is a bunch of noise.

I've been experimenting with only showing the file when it *changed*